storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3088,6 +3119,690 @@ std::ostream &fseg_header::to_stream(std::ostream &out) const
+  {
+    offsets= rec_get_offsets(rec, m_index, offsets, 0,
+                             ULINT_UNDEFINED, &heap);
+    ulint len;
+    byte *field= rec_get_nth_field(rec, offsets,
+                                   rec_offs_n_fields(offsets) - 1, &len);
+    ut_ad(len == 4);
+
+    /* m_parent_pages[child_page_no] =
+       1st 32 bit to indicate offset in parent page
+       2nd 32 bit to indicate parent page number */
+    m_parent_pages[mach_read_from_4(field)]=
+      uint64_t(page_offset(field)) << 32 | block->page.id().page_no();
+    rec= page_rec_get_next(rec);
+  }
+
+  if (heap)
+    mem_heap_free(heap);
+}
+
+uint32_t indexDefragmenter::get_level_block(uint16_t level, dberr_t *err)
+{
+  mtr_t mtr;
+  mtr.start();
+  mtr_x_lock_index(m_index, &mtr);
+  uint32_t child_page_no= m_index->page;
+  mem_heap_t *heap= nullptr;
+  rec_offs *offsets= nullptr;
+  while (1)
+  {
+    dberr_t err= DB_SUCCESS;
+    buf_block_t *block= buf_page_get_gen(page_id_t{0, child_page_no},
+                                         0, RW_S_LATCH, nullptr,
+                                         BUF_GET_POSSIBLY_FREED, &mtr,
+                                         &err);
+    if (!block)
+    {
+      child_page_no= FIL_NULL;
+      break;
+    }
+
+    page_t *page= buf_block_get_frame(block);
+    uint16_t cur_level= btr_page_get_level(page);
+    if (cur_level == level)
+      break;
+    rec_t *rec= page_rec_get_next(page_get_infimum_rec(page));
+    if (rec && !page_rec_is_supremum(rec))
+    {
+      offsets= rec_get_offsets(rec, m_index, offsets,
+                               page_is_leaf(page)
+                               ? m_index->n_core_fields : 0,
+                               ULINT_UNDEFINED, &heap);
+      child_page_no= btr_node_ptr_get_child_page_no(rec, offsets);
+    }
+    else
+    {
+      child_page_no= FIL_NULL;
+      err= DB_CORRUPTION;
+      break;
+    }
+    if (cur_level == level + 1)
+      break;
+  }
+  mtr.commit();
+
+  if (heap)
+    mem_heap_free(heap);
+  return child_page_no;
+}
+
+dberr_t indexDefragmenter::defragment_level(uint16_t level,
+                                            spaceDefragmenter *space_defrag)
+{
+  dberr_t err= DB_SUCCESS;
+  uint32_t cur_page_no= get_level_block(level, &err);
+  if (cur_page_no == FIL_NULL)
+    return err;
+
+  fil_space_t *space= fil_system.sys_space;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  mtr_x_lock_index(m_index, &mtr);
+
+  buf_block_t *block= buf_page_get_gen(page_id_t{0, cur_page_no},
+                                       0, RW_X_LATCH, nullptr,
+                                       BUF_GET_POSSIBLY_FREED, &mtr, &err);
+  if (!block)
+  {
+err_exit:
+    mtr.discard_modifications();
+    mtr.commit();
+    return err;
+  }
+
+  while (block)
+  {
+    page_t *page= buf_block_get_frame(block);
+    uint32_t next_page_no= btr_page_get_next(page);
+    uint32_t cur_extent= (cur_page_no / FSP_EXTENT_SIZE) * FSP_EXTENT_SIZE;
```

Comment:
`FSP_EXTENT_SIZE` is a complex macro. It must not be evaluated several times per loop iteration.